app/src/main/java/org/p2p/wallet/common/ui/widget/EarnWidget.kt

```diff
@@ -0,0 +1,183 @@
+import org.p2p.uikit.utils.getColor
+import org.p2p.uikit.utils.toPx
+import org.p2p.wallet.R
+import org.p2p.wallet.databinding.ItemDepositTokenBinding
+import org.p2p.wallet.databinding.WidgetEarnViewBinding
+import org.p2p.wallet.utils.DecimalFormatter
+import org.p2p.wallet.utils.isZero
+import org.p2p.wallet.utils.viewbinding.inflateViewBinding
+import java.math.BigDecimal
+
+private const val DEFAULT_BUTTON_TEXT_PADDING_DP = 20
+private const val SMALL_BUTTON_TEXT_PADDING_DP = 16
+
+private const val TOKEN_MARGIN_END_DP = 8
+
+private const val TICKER_DECIMALS = 12
+
+class EarnWidget @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyleAttr: Int = 0
+) : FrameLayout(context, attrs, defStyleAttr), KoinComponent {
+
+    private val binding: WidgetEarnViewBinding = inflateViewBinding()
+
+    private val glideManager: GlideManager by inject()
+
+    var currentState: EarnWidgetState? = null
+
+    init {
+        binding.tickerViewAmount.apply {
+            setCharacterLists(TickerUtils.provideNumberList())
+            setPreferredScrollingDirection(TickerView.ScrollingDirection.DOWN)
+        }
+    }
+
+    fun setOnButtonClickListener(buttonClickListener: OnClickListener) {
+        binding.buttonEarn.setOnClickListener(buttonClickListener)
+    }
+
+    fun setWidgetState(state: EarnWidgetState) = with(binding) {
+        currentState = state
+        viewEarnContent.setBackgroundColor(getColor(state.backgroundColor))
+        val balanceState = state is EarnWidgetState.Balance
+        val depositingState = state is EarnWidgetState.Depositing
+        textViewEarnTitle.isVisible = !depositingState
+        tickerViewAmount.isVisible = balanceState
+        textViewEarnMessage.isVisible = !balanceState
+        viewTokenContainer.isVisible = balanceState
+        buttonEarn.isEnabled = !depositingState
+        when (state) {
+            is EarnWidgetState.Balance -> {
+                makeAlignStartContent()
+                tickerViewAmount.text = "$${state.amount.formatTicker()}"
+                textViewEarnTitle.setText(R.string.earn_widget_balance_title)
+                buttonEarn.setText(R.string.earn_widget_balance_button)
+            }
+            is EarnWidgetState.LearnMore -> {
+                makeAlignCenterContent()
+                textViewEarnTitle.setText(R.string.earn_widget_learn_more_title)
+                textViewEarnMessage.setText(R.string.earn_widget_learn_more_message)
+                buttonEarn.setText(R.string.earn_widget_learn_more_button)
+            }
+            is EarnWidgetState.Depositing -> {
+                makeAlignStartContent()
+                tickerViewAmount.isVisible = false
+                textViewEarnTitle.isVisible = false
+                textViewEarnMessage.isVisible = false
+                buttonEarn.setText(state.buttonTextRes)
+                buttonEarn.isEnabled = true // TODO remove after tests and real integration
```

Comment:
should we create a task special for that?

Reply:
no this is only for tests QA all states
any what after implementing logic of widget this should be deleted